### ISSUE TITLE
(+) plus symbol in query parameter turns into space instead of encoding as %2B

### ISSUE THREAD

**Issue by abdulmahamaliyev** [usebruno/bruno#3187]
### I have checked the following:

- [X] I use the newest version of bruno.
- [X] I've searched existing issues and found nothing related to my issue.

### Describe the bug

Updated bruno recently and some of old requests don't work because + sign in query parameter gets replaced by a space.

As a workaround, replacing (+) sign with %2B resolves the issue. 

I'm at work so can't provide screenshot right now

### Screenshots/Live demo link

N/A

**Comment by mroderick** [comment]
Could you add some examples?

**Comment by alifemove** [comment]
I went through this elsewhere, but its not a bug spaces can be interpreted as + or %20.
https://en.wikipedia.org/wiki/Query_string

**Comment by Pragadesh-45** [comment]
> Could you provide some examples?  

Hey @mroderick,  
Here is an example that demonstrates this behavior:

https://github.com/user-attachments/assets/799da1d8-f158-4101-a727-c668c4c7703f

As @alifemove mentioned, this is not a bug. A query string can contain either `+` or `%20` to represent spaces.

**Comment by Dunky13** [comment]
I faced this bug myself today, and was thoroughly confused by it. Thought it was a bug in my code, rather than bruno.

A solution would be:
- Anything that's input in the param is treated as if it wasn't encoded yet, so if you write `+` in the param it will be converted to `%2B`
- Anything that's input in the url is treated as if it was already encoded. So if you write `+` in the url, it will be converted to `%20` or ` ` in the param field.

But when I input a string in the param, I expect it to be treated as a string that will be encoded still, instead of verbatim copying it to the url

**Comment by alifemove** [comment]
As this is not a bug this should be closed. 
The standards say that + is to be treated as a space, which is what Bruno is doing. Any workarounds being created or changes to Bruno would be going against the standards. Standards also say if you want a literal + you should use %2B, which was stated in the original issue that %2B works, because again, it's working as it should.

**Comment by Dunky13** [comment]
@alifemove I agree that the `+` in the url should be treated as a space, as is the spec (and Bruno)

But there is also part of user expectation, that when they (me included) put a + in the param value field, to be treated as a string literal, and be encoded as such. 
Hence, in my first comment [here](https://github.com/usebruno/bruno/issues/3187#issuecomment-2473652685) I suggested that the `+` in the url bar would be treated according to the spec as a whitespace, but the `+` in a param field, would be treated as string literal, and to be encoded.
Since (I) as a user it's not unreasonable to expect that input fields are as seen. - not everyone is fully aware of the intricacies of the spec.
Hope that makes sense

**Comment by drwelby** [comment]
Following up to https://github.com/usebruno/bruno/issues/3187#issuecomment-2473690689, the script should split only the first occurrence of the `=` in the event you have parameters like `&filter=id=1234`.

```
const [url,params] = req.url.split('?');
if(!params) return;
const encodedParams = params.split('&').map((kv) => {
  const [k,...vs] = kv.split('=');
  const v = vs.join("=")
  return `${k}=${encodeURIComponent(v)}`;
}).join('&');
req.setUrl(`${url}?${encodedParams}`);
```

**Comment by Dunky13** [comment]
```const [url,params] = req.url.split('?');
if(!params) return;
const encodedParams = params.split('&').map((kv) => {
  const [k,v] = kv.split('=', 2);
  return `${k}=${encodeURIComponent(v)}`;
}).join('&');
req.setUrl(`${url}?${encodedParams}`);```

Slightly more optimized, split can set a [limit](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#limit)

**Comment by drwelby** [comment]
> Good point! Here's a slightly more optimized, split can set a [limit](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#limit)

Limit works differently than you'd expect (unless you're coming from Java?)

    [k,v] = "filter=a=1 and b=2".split('=', 2);

will result in a `k` = `"filter"` and `v` = `a`

**Comment by Dunky13** [comment]
> > Good point! Here's a slightly more optimized, split can set a [limit](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#limit)
> 
> Limit works differently than you'd expect (unless you're coming from Java?)
> 
> ```
> [k,v] = "filter=a=1 and b=2".split('=', 2);
> ```
> 
> will result in a `k` = `"filter"` and `v` = `a`

Damn you're right. Scratch my previous suggestion!

**Comment by xidsyed** [comment]
ran into the same issue. its sad that the features hasn't been added yet :(

**Comment by SAIF-git903** [comment]
You can actually use [encodeURIComponent ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) function from js

```
const email = 'test+1@gmail.com'
const encodedEmail = encodeURIComponent(email); // Encode to keep the '+'
const path = `/${path}?email=${encodedEmail}`
```

**Comment by loi-avela** [comment]
My use case is I want to accept a query param like an email `some+body@example.com`
I expect the client code to encode the email param like this:
```
/api/users?email=some%2Bbody%40example.com
```
When the user route receives the request. 

```
c.req.valid("query").email => "some body@email.com"
```
I am expecting the email query to be exactly `some+body@example.com`

Without a fix, what would the workaround be?
I have no idea, where I'm supposed to put this code that would allow me to overwrite the query: https://github.com/usebruno/bruno/issues/3187#issuecomment-2529007667

I'm using OpenApiHono v4

**Comment by maintainer-bruno** [comment]
This issue has been resolved by #5089 and is available starting from v2.8.0.

URL encoding is no longer visibly applied in the UI — both the Params table and the Request URL field now display values exactly as entered by the user. Bruno will continue to send encoded values to the API server (unless URL encoding is explicitly disabled in the Settings tab for each request).

> 🔧 Note:
> URL encoding is enabled by default for newly created requests. However, for backward compatibility, requests from older collections will require manual enabling of this setting if needed.

Closing this issue now. Feel free to reopen it if you continue to experience the problem.

**Comment by jonwayne-tithely** [comment]
An option for handling this locally, in a specific `collection.bru` file. This will attempt to decode -> encode every param value so that both encoded and non-encoded ("human readable") param values are ultimately encoded before transmission. 

```
script:pre-request {
  const url = req.getUrl();
  const queryIndex = url.indexOf('?');

  if (queryIndex !== -1) {
    const base = url.slice(0, queryIndex);
    const query = url.slice(queryIndex + 1);

    const encoded = query.split('&').map((pair) => {
      const [key, ...rest] = pair.split('=');
      const value = rest.join('=');

      let decoded = value;
      try {
        decoded = decodeURIComponent(value);
      } catch (e) {
        // not validly percent-encoded (e.g. a bare `%`) — encode as-is
      }

      return `${key}=${encodeURIComponent(decoded)}`;
    }).join('&');

    req.setUrl(`${base}?${encoded}`);
  }
}
```
